File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/DeviceEventSample/src/com/hp/workpath/sample/deviceeventsample/task/DeviceEventTask.java
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.deviceeventsample.task;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import com.hp.workpath.api.Result;
import com.hp.workpath.api.device.events.DeviceEvent;
import com.hp.workpath.api.device.events.DeviceEventsService;
import com.hp.workpath.sample.deviceeventsample.Logger;
import com.hp.workpath.sample.deviceeventsample.fragment.ResponseInterface;

import java.lang.ref.WeakReference;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class DeviceEventTask {

    private WeakReference<Context> mContextRef;
    private ResponseInterface mResponseInterface;
    private Result mResult;

    private ExecutorService executor = Executors.newSingleThreadExecutor();
    private Handler handler = new Handler(Looper.getMainLooper());

    public DeviceEventTask(Context context, ResponseInterface responseInterface) {
        this.mContextRef = new WeakReference<>(context);
        this.mResponseInterface = responseInterface;
        this.mResult = new Result();
    }

    public void taskExecute() {
        try {
            executor.execute(new Runnable() {
                @Override
                public void run() {
                    try {
                        onPostExecute(DeviceEventsService.getDeviceEvents(mContextRef.get(), mResult));
                    } catch (Throwable t) {
                        Logger.showResult(null, "DeviceEventsService.getDeviceEvents " + t.getMessage());
                        executor.shutdown();
                        onPostExecute(null);
                    }
                }
            });
        } catch (Exception e) {
            Logger.showResult(null, "DeviceEventsService.getDeviceEvents " + e.getMessage());
            onPostExecute(null);
            executor.shutdown();
        }
    }

    private void onPostExecute(List<DeviceEvent> deviceEventList) {
        handler.post(new Runnable() {
            @Override
            public void run() {
                if (mResult.getCode() == Result.RESULT_OK) {
                    mResponseInterface.success(deviceEventList);
                } else {
                    mResponseInterface.failure("DeviceEventsService.getDeviceEvents", mResult);
                }
            }
        });
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/DeviceEventSample/src/com/hp/workpath/sample/deviceeventsample/task/DeviceEventTask.kt
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.deviceeventsample.task

import android.content.Context
import com.hp.workpath.api.Result
import com.hp.workpath.api.device.events.DeviceEvent
import com.hp.workpath.api.device.events.DeviceEventsService
import com.hp.workpath.sample.deviceeventsample.fragment.ResponseInterface
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.lang.ref.WeakReference

class DeviceEventTask(context: Context, responseInterface: ResponseInterface) {
    private val mContextRef: WeakReference<Context> = WeakReference(context)
    private val mResponseInterface: ResponseInterface = responseInterface
    private val mResult: Result = Result()
    private var mThrowable: Throwable? = null

    suspend fun execute() {
        var deviceEventList: List<DeviceEvent>? = null
        try {
            mContextRef.get()?.run {
                deviceEventList = DeviceEventsService.getDeviceEvents(this, mResult)
            }
        } catch (t: Throwable) {
            mThrowable = t
        }
        onPostExecute(deviceEventList)
    }

    private suspend fun onPostExecute(deviceEventList: List<DeviceEvent>?) {
        withContext(Dispatchers.Main) {
            if (mResult.code == Result.RESULT_OK && deviceEventList != null) {
                mResponseInterface.success(deviceEventList)
            } else {
                mThrowable?.run {
                    mResponseInterface.failure("DeviceEventsService.getDeviceEvents $message", null)
                } ?: run {
                    mResponseInterface.failure("DeviceEventsService.getDeviceEvents", mResult)
                }
            }
        }
    }
}